<!DOCTYPE html>
<html>

  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="description" content="SMRT-SV : Structural variant and indel caller for PacBio reads">

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">

    <title>SMRT-SV</title>
  </head>

  <body>

    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/EichlerLab/pacbio_variant_caller">View on GitHub</a>

          <h1 id="project_title">SMRT-SV</h1>
          <h2 id="project_tagline">Structural variant and indel caller for PacBio reads</h2>

            <section id="downloads">
              <a class="zip_download_link" href="https://github.com/EichlerLab/pacbio_variant_caller/zipball/master">Download this project as a .zip file</a>
              <a class="tar_download_link" href="https://github.com/EichlerLab/pacbio_variant_caller/tarball/master">Download this project as a tar.gz file</a>
            </section>
        </header>
    </div>

    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">
      <section id="main_content" class="inner">
        <h1>
<a id="smrt-sv" class="anchor" href="#smrt-sv" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>SMRT-SV</h1>

<p>Structural variant (SV) and indel caller for PacBio reads based on methods from
<a href="http://www.nature.com/nature/journal/vaop/ncurrent/full/nature13907.html">Chaisson et
al. 2014</a>.</p>

<h2>
<a id="installation" class="anchor" href="#installation" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Installation</h2>

<p>SMRT-SV requires git, Python (2.6.6 or later) and Perl (5.10.1 or later) for
installation.</p>

<p>SMRT-SV has been tested on CentOS 6.8 and should work with most Linux-style
distributions.</p>

<h3>
<a id="get-the-code" class="anchor" href="#get-the-code" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Get the code</h3>

<p>Clone the repository into your desired installation directory and build SMRT-SV
dependencies.</p>

<div class="highlight highlight-source-shell"><pre>mkdir /usr/local/smrtsv
<span class="pl-c1">cd</span> /usr/local/smrtsv
git clone --recursive git@github.com:EichlerLab/pacbio_variant_caller.git <span class="pl-c1">.</span>
make</pre></div>

<p>Note that some dependencies (e.g., RepeatMasker) require hardcoded paths to this
installation directory. If you need to move SMRT-SV to another directory, it is
easier to change to that directory, clone the repository, and rebuild the
dependencies there.</p>

<h3>
<a id="test-installation" class="anchor" href="#test-installation" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Test installation</h3>

<p>Add the installation directory to your path.</p>

<div class="highlight highlight-source-shell"><pre><span class="pl-k">export</span> PATH=/usr/local/smrtsv:<span class="pl-smi">$PATH</span></pre></div>

<p>Print SMRT-SV help to confirm installation.</p>

<div class="highlight highlight-source-shell"><pre>smrtsv.py --help</pre></div>

<p>Alternately, run <code>smrtsv.py</code> directly from the installation directory.</p>

<div class="highlight highlight-source-shell"><pre>/usr/local/smrtsv/bin/smrtsv.py --help</pre></div>

<h2>
<a id="configure-distributed-environment" class="anchor" href="#configure-distributed-environment" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Configure distributed environment</h2>

<p>SMRT-SV uses DRMAA to submit jobs to a grid-engine-style cluster. To enable the <code>--distribute</code> option of SMRT SV, add the following line to your <code>.bash_profile</code> with the correct path to the DRMAA library for your cluster.</p>

<div class="highlight highlight-source-shell"><pre><span class="pl-k">export</span> DRMAA_LIBRARY_PATH=/opt/uge/lib/lx-amd64/libdrmaa.so.1.0</pre></div>

<p>Alternately, provide the path to your DRMAA library with the SMRT-SV
<code>--drmaalib</code> option.</p>

<p>Additionally, you may need to configure resource requirements depending on your
cluster and PacBio data. Use the <code>--cluster_config</code> option when running SMRT-SV
to pass a JSON file that specifies <a href="https://bitbucket.org/snakemake/snakemake/wiki/Documentation#markdown-header-cluster-configuration">Snakemake-style cluster
parameters</a>. An
example configuration used to run SMRT-SV with human genomes on the Eichler lab
cluster is provided in this repository in the file <code>cluster.eichler.json</code>.</p>

<h2>
<a id="tutorial" class="anchor" href="#tutorial" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Tutorial</h2>

<p>The following tutorial shows how to call structural variants and indels in
yeast.</p>

<h3>
<a id="download-pacbio-reads" class="anchor" href="#download-pacbio-reads" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Download PacBio reads</h3>

<div class="highlight highlight-source-shell"><pre><span class="pl-c"># List of AWS-hosted files from PacBio including raw reads and an HGAP assembly.</span>
wget https://gist.githubusercontent.com/pb-jchin/6359919/raw/9c172c7ff7cbc0193ce89e715215ce912f3f30e6/gistfile1.txt

<span class="pl-c"># Keep only .xml, .bas.h5, and .bax.h5 files.</span>
sed <span class="pl-s"><span class="pl-pds">'</span>/fasta/d;/fastq/d;/celera/d;/HGAP/d<span class="pl-pds">'</span></span> gistfile1.txt <span class="pl-k">&gt;</span> gistfile1.keep.txt

<span class="pl-c"># Download data into a raw reads directory.</span>
mkdir -p raw_reads
<span class="pl-c1">cd</span> raw_reads
<span class="pl-k">for</span> <span class="pl-smi">f</span> <span class="pl-k">in</span> <span class="pl-s"><span class="pl-pds">`</span>cat ../gistfile1.keep.txt<span class="pl-pds">`</span></span><span class="pl-k">;</span> <span class="pl-k">do</span> wget --force-directories <span class="pl-smi">$f</span><span class="pl-k">;</span> <span class="pl-k">done</span>

<span class="pl-c"># Create a list of reads for analysis.</span>
<span class="pl-c1">cd</span> ..
find ./raw_reads -name <span class="pl-s"><span class="pl-pds">"</span>*.bax.h5<span class="pl-pds">"</span></span> -exec readlink -f {} <span class="pl-cce">\;</span> <span class="pl-k">&gt;</span> reads.fofn</pre></div>

<h3>
<a id="prepare-the-reference-assembly" class="anchor" href="#prepare-the-reference-assembly" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Prepare the reference assembly</h3>

<p>Download the reference assembly (sacCer3) from UCSC.</p>

<div class="highlight highlight-source-shell"><pre>mkdir -p reference
<span class="pl-c1">cd</span> reference
wget ftp://hgdownload.cse.ucsc.edu/goldenPath/sacCer3/bigZips/chromFa.tar.gz</pre></div>

<p>Unpack the reference tarball and concatenate individual chromosome files into a
single reference FASTA file.</p>

<div class="highlight highlight-source-shell"><pre>tar zxvf chromFa.tar.gz
cat <span class="pl-k">*</span>.fa <span class="pl-k">&gt;</span> sacCer3.fasta
rm -f <span class="pl-k">*</span>.fa <span class="pl-k">*</span>.gz
<span class="pl-c1">cd</span> ..</pre></div>

<p>Prepare the reference sequence for alignment with PacBio reads. This step
produces suffix array and ctab files used by BLASR to speed up alignments.</p>

<div class="highlight highlight-source-shell"><pre>smrtsv.py index reference/sacCer3.fasta</pre></div>

<h3>
<a id="align-reads-to-the-reference" class="anchor" href="#align-reads-to-the-reference" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Align reads to the reference</h3>

<p>Align reads to the reference with BLASR.</p>

<div class="highlight highlight-source-shell"><pre>smrtsv.py align reference/sacCer3.fasta reads.fofn</pre></div>

<h3>
<a id="find-signatures-of-variants-in-raw-reads" class="anchor" href="#find-signatures-of-variants-in-raw-reads" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Find signatures of variants in raw reads</h3>

<p>Find candidate regions to search for SVs based on SV signatures.</p>

<div class="highlight highlight-source-shell"><pre>smrtsv.py detect reference/sacCer3.fasta alignments.fofn candidates.bed</pre></div>

<h3>
<a id="assemble-regions" class="anchor" href="#assemble-regions" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Assemble regions</h3>

<p>Assemble local regions of the genome that have SV signatures or tile across the
genome.</p>

<div class="highlight highlight-source-shell"><pre>smrtsv.py assemble reference/sacCer3.fasta reads.fofn alignments.fofn candidates.bed local_assembly_alignments.bam</pre></div>

<h3>
<a id="call-variants" class="anchor" href="#call-variants" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Call variants</h3>

<p>Call variants by aligning tiled local assemblies back to the
reference. Optionally, specify the sample name for annotation of the final VCF
file and a species name (common or scientific as supported by
<a href="http://www.repeatmasker.org/">RepeatMasker</a>) for repeat masking of structural
variants.</p>

<div class="highlight highlight-source-shell"><pre>smrtsv.py call reference/sacCer3.fasta alignments.fofn local_assembly_alignments.bam variants.vcf --sample UCSF_Yeast9464 --species yeast</pre></div>
      </section>
    </div>

    <!-- FOOTER  -->
    <div id="footer_wrap" class="outer">
      <footer class="inner">
        <p class="copyright">SMRT-SV maintained by <a href="https://github.com/EichlerLab">EichlerLab</a></p>
        <p>Published with <a href="https://pages.github.com">GitHub Pages</a></p>
      </footer>
    </div>

    

  </body>
</html>
