"""
Rules for local assembly of genomic regions.
"""
print("Local assembly with MHAP/Celera")

#
# Inputs:
#  1. Text file with a list of absolute paths for BAMs with reads for assembly
#  2. BED file with a list of regions to assemble
#
# For signature-based SV calling, the list of regions is based on signatures of
# SVs. For the tiled-based SV calling, the regions are sliding windows across
# the entire genome.

# Load regions to assemble.
if config.get("assembly") and config["assembly"].get("regions_to_assemble"):
    with open(config["assembly"]["regions_to_assemble"], "r") as fh:
        REGIONS_TO_ASSEMBLE = ["-".join(line.rstrip().split("\t")[:3]) for line in fh]
else:
    REGIONS_TO_ASSEMBLE = []

#
# Define internal constants.
#
ASSEMBLY_DIR = "mhap_assembly"

#
# Define rules.
#

# Collect assemblies.
rule collect_assemblies:
    input: expand("mhap_assembly/{region}/consensus.fasta", region=REGIONS_TO_ASSEMBLE)
    output: "sv_assemblies.txt"
    params: sge_opts=""
    # TODO: create module for Celera scripts
    shell: "find {input} > {output}"

# TODO: remove dependencies on mchaisso's scripts (*.py) and Celera installation.
# TODO: remove dependencies on Eichler modules.
rule assemble_region:
    input: "alignments.fofn"
    output: "{ASSEMBLY_DIR}/{region}/consensus.fasta"
    # Limit run time of the assembly to 2 hours to prevent long-running assembly
    # of overly difficult regions.
    params: sge_opts="-l mfree=2G -pe serial 4 -l disk_free=10G -l h_rt=02:00:00"
    shell:
        "export ANALYSIS_DIR=`pwd`; "
        "mkdir -p {TMP_DIR}/{wildcards.region}; "
        "pushd {TMP_DIR}/{wildcards.region}; "
        "rsync $ANALYSIS_DIR/config.json {TMP_DIR}/{wildcards.region}/; "
        "snakemake -s $ANALYSIS_DIR/rules/local_assembly.mhap_celera_single_assembly.rules --config alignments=$ANALYSIS_DIR/{input} region={wildcards.region}; "
        "popd; "
        "rsync -ra {TMP_DIR}/{wildcards.region} {ASSEMBLY_DIR}/; "
        "rm -rf {TMP_DIR}/{wildcards.region}"
